### ISSUE TITLE
reactive query parameters with built-in `useAsyncData`

### ISSUE THREAD

**Issue by alexanderivn** [Intevel/nuxt-directus#238]
Hello, i have question regarding of watch and search inside the module, im trying to implement search function and filtering function, but it seems doesnt work, the api call from search value is always empty, this is working fine if i use the nuxt-directus plugin (https://github.com/Intevel/nuxt-directus).

here is my composables.
```
export default async function useGetAllServices(): Promise<{
  refresh: unknown
  siteUrl: unknown
  pending: unknown
  error: unknown
  search: unknown
  services: Services[]
}> {
  const { readItems } = useDirectusItems()
  const { client: siteUrl } = useDirectusFiles()
  const search = ref('')

  const {
    data: services,
    error,
    pending,
    refresh,
  } = await readItems('services', {
    key: 'service',
    params: {
      watch: [search],
      lazy: true,
      immediate: true,
    },
    query: {
      fields: [
        'title',
        'slug',
        'icon',
        'small_description',
        'content',
        { icon: ['id', 'filename_disk', 'filename_download'] },
      ],
      search: search.value,
      filter: {
        // @ts-ignore
        status: {
          _eq: 'published',
        },
      },
    },
  })
  return { services, error, pending, refresh, siteUrl, search }
}
```

And this is my vue component
```
<script lang="ts" setup>
const { services, error, refresh, pending, siteUrl, search } = await useGetAllServices()
</script>

<template>
<section class="container mx-auto">
     <input v-model="search" class="my-8 w-full border py-4" type="text" />
     .......
</section>
</template>
```

but the search response value always empty.

![firefox_yIZVHvwbgi](https://github.com/Intevel/nuxt-directus/assets/13149076/3bc1cacb-3a5b-42cb-80b9-6bd7fb3b7c00)

Am i doing something wrong here? thank you!

**Comment by Sandros94** [comment]
>here is my composables.
>```ts
>// [...]
>const { client: siteUrl } = useDirectusFiles()
>// [...]
>```
First thing first, just to make sure, this isn't the correct way to get your directus `siteUrl`. That `client` is used internally by the module or for more advanced use cases using the SDK directly (you can read about it in the [official Directus's Docs](https://docs.directus.io/guides/sdk/getting-started.html)):

If you configured your `directus.url` correctly in `nuxt.config.ts`, you can access it via:
```ts
const { url } = useRuntimeConfig().publlic.directus
```

or if you would like to change the url only for that particular composable you could do this:
```ts
const { readItems } = useDirectusItems({ options: { baseURL: 'https://my-different-directus-instance.com' } })
```

But more importantly, could you share a copy of the error you are getting in the console, or better of the `error` export of your composable?

**Comment by alexanderivn** [comment]
Thank you for replying.

ah. i use this to get the image
```const { client: siteUrl } = useDirectusFiles()```

```:src="`${siteUrl.url}assets/${service.icon?.filename_disk}?key=webp-icon`"```.

for my issue, there is no error in the console, i can get the data and show it to the front end without a problem, but when i try to search and use watch, its just return an empty search value. like shown below

![image](https://github.com/Intevel/nuxt-directus/assets/13149076/9ebb98b3-cd3e-46b6-b8ce-72570c7bff5f)

Thank you.

**Comment by Sandros94** [comment]
> vue component
> `` :src="`${siteUrl.url}assets/${service.icon?.filename_disk}?key=webp-icon`" ``.

oh, that is an interesting use and never realized that. Perhaps you could export the url from your `useDirectusItems()` that you are already using, instead of calling yet another composable+client. Better of, exposing only the url itself:
```ts
const { readItems, client: { url } } = useDirectusItems()
```

On the main topic, sorry for the distraction:

I'm testing this right now, and other than noticing an error with how I implemented `useAsyncData` I'm also able to reproduce the issue. It might not be that simple to solve.

**Comment by Sandros94** [comment]
I'm going to tag the boys here: @Intevel @casualmatt

This is my simple reproduction in our playground:

### [`~/playground/composables/test.ts`](https://github.com/Intevel/nuxt-directus/blob/2cbf154ae52b90ba5d561757db8e81fb579c11a0/playground/composables/test.ts)
```ts
import type { Ref } from 'vue'
import { readItems as readItemsSDK } from '@directus/sdk'
import { useAsyncData, useDirectusItems } from '#imports'
import type { Schema } from '../types'

export function myComposable () {
  const { client } = useDirectusItems<Schema>()

  async function readItemsTest (fieldParam: Ref<string>, searchParam: Ref<string>) {
    return await useAsyncData(
      () => client.request(readItemsSDK('posts', {
        fields: [fieldParam.value],
        search: searchParam.value
      })), {
        immediate: false,
        watch: [fieldParam, searchParam]
      }
    )
  }

  return {
    readItemsTest
  }
}
```

### [`~/playground/pages/test.vue`](https://github.com/Intevel/nuxt-directus/blob/2cbf154ae52b90ba5d561757db8e81fb579c11a0/playground/pages/test.vue)
```vue
<template>
  <div>
    <input v-model="fieldParam" type="text" placeholder="Field Param">
    <input v-model="searchParam" type="text" placeholder="Search posts">
    <button @click="refresh(); refreshToo(); refreshThree()">
      Refresh
    </button>
    <pre v-if="notWorking && notWorking.length > 0">
      useAsyncData
      {{ notWorking }}
    </pre>
    <pre v-if="working && working.length > 0">
      SDK
      {{ working }}
    </pre>
    <pre v-if="workingToo && workingToo.length > 0">
      myComposable
      {{ workingToo }}
    </pre>
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue'
import { readItems as readItemsSDK } from '@directus/sdk'
import { useAsyncData, useDirectusItems } from '#imports'
import { myComposable } from '../composables/test'
import type { Schema } from '../types'

const { readItems, client } = useDirectusItems<Schema>()

const fieldParam = ref<string>('id')
const searchParam = ref<string>('')

const { data: notWorking, refresh } = await readItems('posts', {
  query: {
    fields: [fieldParam.value],
    search: searchParam.value
  },
  params: {
    immediate: false,
    watch: [fieldParam, searchParam]
  }
})

const { data: working, refresh: refreshToo } = await useAsyncData(
  () => client.request(readItemsSDK('posts', {
    fields: [fieldParam.value],
    search: searchParam.value
  })), {
    immediate: false,
    watch: [fieldParam, searchParam]
  }
)

const { readItemsTest } = myComposable()

const { data: workingToo, refresh: refreshThree } =
  await readItemsTest(fieldParam, searchParam)
</script>
```

In short: something strange is happening with `useAsyncData` when it is part of a module. From what it seems it bakes the query parameters into the sdk function, as shown here:

https://github.com/Intevel/nuxt-directus/assets/13056429/ceb50106-6541-47fc-b9e8-e33671128495

Should we ask the bigger boys?

**Comment by Sandros94** [comment]
Oh, now that I think about it, this also explains a few issue I had months ago about `collection`s variables and though were solved placing them into `ref`s

**Comment by Sandros94** [comment]
in the meanwhile @alexanderivn you could create your local composable like this:

```js
import { readItems } from '@directus/sdk'

export default async function useGetAllServices () {
  const { client } = useDirectusRest()
  const search = ref('')

  return {
    url: client.url,
    search,
    services: await useAsyncData('services', () =>
      client.request(readItems('services', {
        fields: [
          'title',
          'slug',
          'icon',
          'small_description',
          'content',
          { icon: ['id', 'filename_disk', 'filename_download'] },
        ],
        search: search.value,
        filter: {
          status: {
            _eq: 'published',
          },
        },
      }), {
        watch: [search],
        lazy: true,
        immediate: true,
      }))
  }
}
```

**Comment by alexanderivn** [comment]
> in the meanwhile @alexanderivn you could create your local composable like this:
> 
> ```js
> import { readItems } from '@directus/sdk'
> 
> export default async function useGetAllServices () {
>   const { client } = useDirectusRest()
>   const search = ref('')
> 
>   return {
>     url: client.url,
>     search,
>     services: await useAsyncData('services', () =>
>       client.request(readItems('services', {
>         fields: [
>           'title',
>           'slug',
>           'icon',
>           'small_description',
>           'content',
>           { icon: ['id', 'filename_disk', 'filename_download'] },
>         ],
>         search: search.value,
>         filter: {
>           status: {
>             _eq: 'published',
>           },
>         },
>       }), {
>         watch: [search],
>         lazy: true,
>         immediate: true,
>       }))
>   }
> }
> ```

Thank you, Sorry for the late reply, yes will try this.. 

Thank you so much!

**Comment by Sandros94** [comment]
I might have understood what is going on.

In short Vue loses the reactivity tracking for things inside non-reactive parameters like `params.query`.

I've noticed this while fixing the reactivity for the collection names in https://github.com/Intevel/nuxt-directus/commit/23f1befdb2bd66ea6194c2f16d2bdbe7931cdfdb, that was a workaround that I though I could use while first rewriting this module, but didn't fully understand why it was needed.

If my understanding on Vue's reactivity system is correct: it is like putting an outter shell to the value you want to make reactive and it acts like a fingerprint. Once Vue starts, it keeps a record for only those fingerprints (I imagine for performance reasons), if the value changes the outter shell updates its fingerprint. Vue, on the next tick of operations, while going through the code it recognizes the reactive shell, checks its fingerprint with the one that already has and if it is different that re-elaborate all the functions that depends from that value.

If my understanding is correct, we might have an hard time using `useAsyncData` within our module...

**Comment by Sandros94** [comment]
### FYI

I'm delaying this till later we are in a stable `6.0.0` version, since it will require quite some time to fix.

**Comment by casualmatt** [comment]
Update:
UseAsyncData will not be implemented, and we change direction for the Nuxt-Directus composable. 
See: https://github.com/Intevel/nuxt-directus/discussions/215#discussioncomment-8404969

**Comment by Sandros94** [comment]
I'm reopening this to mainly track reactivity of query parameters for the new implementation of `useAsyncData`.

More info can be found in the release discussed at [#215 (comment)](https://github.com/Intevel/nuxt-directus/discussions/215#discussioncomment-8574579).

**Comment by casualmatt** [comment]
--> https://github.com/Intevel/nuxt-directus/discussions/271

nuxt-directus-next is EOL

And the new v6 will have first support for `useAsyncData`
